Multiple register in ZPoller fails

I tried to do multi-register of call back in ZPoller, each time with a different context.

So I wrote the following test code:

```java
    @Test
    public void testChannel() throws IOException
    {
        Selector s = Selector.open();
        AtomicInteger count = new AtomicInteger();
        AtomicInteger error = new AtomicInteger();

        EventsHandler cb1 = new EventsHandler() {

            @Override
            public boolean events(Socket socket, int events) {
                throw new UnsupportedOperationException("Not registred for Socket");
            }

            @Override
            public boolean events(SelectableChannel channel, int events) {
                assert (events & ZPoller.IN) > 0;
                try {
                    Pipe.SourceChannel sc = (Pipe.SourceChannel) channel;
                    sc.read(ByteBuffer.allocate(5));
                    count.incrementAndGet();
                    return true;
                } catch (IOException e) {
                    return false;
                }
            }

        };
        EventsHandler cb2 = new EventsHandler() {

            @Override
            public boolean events(Socket socket, int events) {
                throw new UnsupportedOperationException("Not registred for Socket");
            }

            @Override
            public boolean events(SelectableChannel channel, int events) {
                assert (events & ZPoller.IN) > 0;
                try {
                    Pipe.SourceChannel sc = (Pipe.SourceChannel) channel;
                    sc.read(ByteBuffer.allocate(5));
                    count.incrementAndGet();
                    return true;
                } catch (IOException e) {
                    return false;
                }
            }

        };
        EventsHandler cberr = new EventsHandler() {

            @Override
            public boolean events(Socket socket, int events) {
                throw new UnsupportedOperationException("Not registred for Socket");
            }

            @Override
            public boolean events(SelectableChannel channel, int events) {
                error.incrementAndGet();
                return false;
            }

        };

        try (ZPoller poller = new ZPoller(s)) {
            Pipe pipe1 = Pipe.open();
            Pipe.SourceChannel source1 = pipe1.source();
            source1.configureBlocking(false);
            poller.register(source1, cb1, ZPoller.IN);
            poller.register(source1, cb2, ZPoller.IN);
            poller.register(source1, cberr, ZPoller.ERR);

            Pipe pipe2 = Pipe.open();
            Pipe.SourceChannel source2 = pipe2.source();
            source2.configureBlocking(false);
            poller.register(source2, cb1, ZPoller.IN);
            poller.register(source2, cb2, ZPoller.IN);
            poller.register(source2, cberr, ZPoller.ERR);

            pipe1.sink().write(ByteBuffer.wrap(new byte[] {1,2,3}));
            pipe2.sink().write(ByteBuffer.wrap(new byte[] {1,2,3}));
            while (poller.poll(100) > 0) {
                
            }
            assertThat("Not all events handlers checked", count.get(), is(equalTo(2)));
            assertThat("Not all events handlers checked", error.get(), is(equalTo(0)));
        }
    }
```
It creates two generic call back, and one error callback.
It creates two pair of SelectableChannel, register each call back for each and wrote data in each.
The AtomicInteger count check how many time at least one of the generic callback was called.

We can expect a value of :
 - 2 once data was read in a channel, the next callback is not called in the same poll.
 - 4 once data was read in a channel, the next callback is still read during the same poll, even if there is not any more data to read.

But the test fails with a value of 1. The least expected value.

Am I doing something wrong ? Or is that a bug ?

If the last register for error override all registration for the channel, why I didn't get 0 ? I have two channel with exactly the same registration. So I should only get an even value.

More, I have another problem.

If I run this code from eclipse, it fails:
```
java.lang.AssertionError: Not all events handlers checked
Expected: is <2>
     but: was <1>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:865)
	at org.zeromq.TestZPoller.testChannel(TestZPoller.java:477)
....
```
But I got success from maven:
```
$ JAVA_HOME=$(/usr/libexec/java_home -v 1.8) mvn clean package site
...
Results :

Tests run: 491, Failures: 0, Errors: 0, Skipped: 14
...
$ cat target/surefire-reports/TEST-org.zeromq.TestZPoller.xml
<testsuite tests="12" failures="0" name="org.zeromq.TestZPoller" time="0.141" errors="0" skipped="0">
...
  <testcase classname="org.zeromq.TestZPoller" name="testChannel" time="0.103"/>
...
``

--- a/src/main/java/org/zeromq/ZPoller.java
+++ b/src/main/java/org/zeromq/ZPoller.java
@@ -10,11 +10,14 @@ import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 
 import org.zeromq.ZMQ.Poller;
 import org.zeromq.ZMQ.Socket;
+
+import zmq.poll.PollItem;
 
 /**
  * Rewritten poller for 0MQ.
@@ -100,7 +103,7 @@ public class ZPoller implements Closeable
     public interface ItemHolder
     {
         // the inner ZMQ poll item
-        zmq.poll.PollItem item();
+        PollItem item();
 
         // the related ZeroMQ socket
         Socket socket();
@@ -153,7 +156,7 @@ public class ZPoller implements Closeable
         }
 
         @Override
-        public zmq.poll.PollItem item()
+        public PollItem item()
         {
             return base();
         }
@@ -225,6 +228,91 @@ public class ZPoller implements Closeable
         public EventsHandler handler()
         {
             return handler;
+        }
+    }
+
+    private static class CompositePollItem implements ItemHolder, EventsHandler
+    {
+        private final Collection<ItemHolder> holders = new HashSet<>();
+        private final Socket                 socket;
+        private final SelectableChannel      channel;
+
+        private PollItem      item;
+        private EventsHandler globalHandler;
+
+        public CompositePollItem(final Object socketOrChannel)
+        {
+            this.socket = socketOrChannel instanceof Socket ? (Socket) socketOrChannel : null;
+            this.channel = socketOrChannel instanceof SelectableChannel ? (SelectableChannel) socketOrChannel : null;
+            assert (socket != null || channel != null);
+        }
+
+        @Override
+        public PollItem item()
+        {
+            if (item == null) {
+                item = createItem();
+            }
+            return item;
+        }
+
+        private PollItem createItem()
+        {
+            if (socket == null) {
+                return new PollItem(channel, ops());
+            }
+            else {
+                return new PollItem(socket.base(), ops());
+            }
+        }
+
+        private int ops()
+        {
+            return holders.stream().map(holder -> holder.item().zinterestOps()).reduce(this::or).orElse(0);
+        }
+
+        @Override
+        public Socket socket()
+        {
+            return socket;
+        }
+
+        @Override
+        public EventsHandler handler()
+        {
+            return this;
+        }
+
+        @Override
+        public boolean events(Socket socket, int events)
+        {
+            return holders.stream().filter(holder -> holder.item().hasEvent(events))
+                    .map(holder -> holder.handler() == null ? globalHandler : holder.handler()).filter(Objects::nonNull)
+                    .map(handler -> handler.events(socket, events)).reduce(this::and).orElse(false);
+        }
+
+        @Override
+        public boolean events(SelectableChannel channel, int events)
+        {
+            return holders.stream().filter(holder -> holder.item().hasEvent(events))
+                    .map(holder -> holder.handler() == null ? globalHandler : holder.handler()).filter(Objects::nonNull)
+                    .map(handler -> handler.events(channel, events)).reduce(this::and).orElse(false);
+        }
+
+        private Boolean and(Boolean b0, Boolean b1)
+        {
+            return b0 & b1;
+        }
+
+        private Integer or(Integer ops1, Integer ops2)
+        {
+            return ops1 | ops2;
+        }
+
+        private ItemHolder handler(EventsHandler handler)
+        {
+            globalHandler = handler;
+            return this;
         }
     }
 
@@ -334,7 +422,7 @@ public class ZPoller implements Closeable
         this.creator = creator;
         this.selector = selector;
         items = new HashMap<>();
-        all = createContainer(0);
+        all = new HashSet<>();
     }
 
     // creates a new poll item
@@ -446,10 +534,10 @@ public class ZPoller implements Closeable
         if (socketOrChannel == null) {
             return false;
         }
-        Set<ItemHolder> items = this.items.remove(socketOrChannel);
+        CompositePollItem items = this.items.remove(socketOrChannel);
         boolean rc = items != null;
         if (rc) {
-            all.removeAll(items);
+            all.remove(items);
         }
         return rc;
     }
@@ -488,7 +576,7 @@ public class ZPoller implements Closeable
     protected int poll(final long timeout, final boolean dispatchEvents)
     {
         // get all the raw items
-        final Set<zmq.poll.PollItem> pollItems = all.stream().map(ItemHolder::item).collect(Collectors.toSet());
+        final Set<PollItem> pollItems = all.stream().map(ItemHolder::item).collect(Collectors.toSet());
         // polling time
         final int rc = poll(selector, timeout, pollItems);
 
@@ -505,12 +593,16 @@ public class ZPoller implements Closeable
         return -1;
     }
 
-    // does the effective polling
+    private boolean dispatch(Set<CompositePollItem> all, int size)
+    {
+        return dispatch(all.stream().map(aggr -> aggr.handler(globalHandler)).collect(Collectors.toList()), size);
+    }
 
+    // does the effective polling
     protected int poll(final Selector selector, final long tout, final Collection<zmq.poll.PollItem> items)
     {
         final int size = items.size();
-        return zmq.ZMQ.poll(selector, items.toArray(new zmq.poll.PollItem[size]), size, tout);
+        return zmq.ZMQ.poll(selector, items.toArray(new PollItem[size]), size, tout);
     }
 
     /**
@@ -533,7 +625,7 @@ public class ZPoller implements Closeable
                 // no handler, short-circuit
                 continue;
             }
-            final zmq.poll.PollItem item = holder.item();
+            final PollItem item = holder.item();
             final int events = item.readyOps();
 
             if (events <= 0) {
@@ -606,7 +698,7 @@ public class ZPoller implements Closeable
     // checks for read event
     public boolean readable(final Object socketOrChannel)
     {
-        final zmq.poll.PollItem it = filter(socketOrChannel, READABLE);
+        final PollItem it = filter(socketOrChannel, READABLE);
         if (it == null) {
             return false;
         }
@@ -658,7 +750,7 @@ public class ZPoller implements Closeable
     // checks for write event
     public boolean writable(final Object socketOrChannel)
     {
-        final zmq.poll.PollItem it = filter(socketOrChannel, WRITABLE);
+        final PollItem it = filter(socketOrChannel, WRITABLE);
         if (it == null) {
             return false;
         }
@@ -710,7 +802,7 @@ public class ZPoller implements Closeable
     // checks for error event
     public boolean error(final Object socketOrChannel)
     {
-        final zmq.poll.PollItem it = filter(socketOrChannel, ERR);
+        final PollItem it = filter(socketOrChannel, ERR);
         if (it == null) {
             return false;
         }
@@ -750,9 +842,9 @@ public class ZPoller implements Closeable
     // creator of items
     private final ItemCreator creator;
     // managed items
-    private final Map<Object, Set<ItemHolder>> items;
+    private final Map<Object, CompositePollItem> items;
     // all managed items to avoid penalty cost when dispatching
-    private final Set<ItemHolder> all;
+    private final Set<CompositePollItem> all;
 
     // TODO set of handlers, each with its specified events?
     // optional global events handler
@@ -792,15 +884,16 @@ public class ZPoller implements Closeable
         }
         assert (socketOrChannel != null);
 
-        Set<ItemHolder> holders = items.computeIfAbsent(socketOrChannel, i -> createContainer(1));
-        final boolean rc = holders.add(holder);
+        CompositePollItem aggregate = items.computeIfAbsent(socketOrChannel, CompositePollItem::new);
+        final boolean rc = aggregate.holders.add(holder);
         if (rc) {
-            all.add(holder);
+            all.add(aggregate);
         }
         return rc;
     }
 
     // create the container of holders
+    @Deprecated
     protected Set<ItemHolder> createContainer(int size)
     {
         return new HashSet<>(size);
@@ -809,33 +902,23 @@ public class ZPoller implements Closeable
     // gets all the items of this poller
     protected Collection<ItemHolder> items()
     {
-        return all;
+        return all.stream().map(holder -> holder.handler(globalHandler)).collect(Collectors.toSet());
     }
 
     // gets all the items of this poller regarding the given input
     protected Iterable<ItemHolder> items(final Object socketOrChannel)
     {
-        final Set<ItemHolder> set = items.get(socketOrChannel);
-        if (set == null) {
+        final CompositePollItem aggregate = items.get(socketOrChannel);
+        if (aggregate == null) {
             return Collections.emptySet();
         }
-        return set;
+        return aggregate.holders;
     }
 
     // filters items to get the first one matching the criteria, or null if none found
-    protected zmq.poll.PollItem filter(final Object socketOrChannel, int events)
+    protected PollItem filter(final Object socketOrChannel, int events)
     {
-        if (socketOrChannel == null) {
-            return null;
-        }
-
-        final Iterable<ItemHolder> items = items(socketOrChannel);
-        for (ItemHolder item : items) {
-            final zmq.poll.PollItem it = item.item();
-            if (it.hasEvent(events)) {
-                return it;
-            }
-        }
-        return null;
+        return Optional.ofNullable(socketOrChannel).map(items::get).map(CompositePollItem::item)
+                .filter(item -> item.hasEvent(events)).orElse(null);
     }
 }

--- a/src/main/java/zmq/poll/PollItem.java
+++ b/src/main/java/zmq/poll/PollItem.java
@@ -94,7 +94,6 @@ public class PollItem
         return interest;
     }
 
-    @Deprecated
     public int zinterestOps()
     {
         return zinterest;

--- a/src/test/java/org/zeromq/TestZPoller.java
+++ b/src/test/java/org/zeromq/TestZPoller.java
@@ -10,7 +10,11 @@ import static org.junit.Assert.assertThat;
 import static org.junit.Assert.fail;
 
 import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Pipe;
 import java.nio.channels.SelectableChannel;
+import java.nio.channels.Selector;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
 import org.junit.Test;
@@ -21,6 +25,60 @@ import org.zeromq.ZPoller.ItemHolder;
 
 public class TestZPoller
 {
+    private final class EventsHandlerCounter implements EventsHandler
+    {
+        private final AtomicInteger count;
+
+        private EventsHandlerCounter(AtomicInteger count)
+        {
+            this.count = count;
+        }
+
+        @Override
+        public boolean events(Socket socket, int events)
+        {
+            throw new UnsupportedOperationException("Not registred for Socket");
+        }
+
+        @Override
+        public boolean events(SelectableChannel channel, int events)
+        {
+            assertThat((events & ZPoller.IN) > 0, is(true));
+            try {
+                Pipe.SourceChannel sc = (Pipe.SourceChannel) channel;
+                sc.read(ByteBuffer.allocate(5));
+                count.incrementAndGet();
+                return true;
+            }
+            catch (IOException e) {
+                return false;
+            }
+        }
+    }
+
+    private final class EventsHandlerErrorCounter implements EventsHandler
+    {
+        private final AtomicInteger error;
+
+        private EventsHandlerErrorCounter(AtomicInteger error)
+        {
+            this.error = error;
+        }
+
+        @Override
+        public boolean events(Socket socket, int events)
+        {
+            throw new UnsupportedOperationException("Not registred for Socket");
+        }
+
+        @Override
+        public boolean events(SelectableChannel channel, int events)
+        {
+            error.incrementAndGet();
+            return false;
+        }
+    }
+
     private static class EventsHandlerAdapter implements EventsHandler
     {
         @Override
@@ -375,5 +433,55 @@ public class TestZPoller
             poller.close();
             ctx.close();
         }
+    }
+
+    @Test
+    public void testMultipleRegistrations() throws IOException
+    {
+        Selector selector = Selector.open();
+        AtomicInteger count = new AtomicInteger();
+        AtomicInteger error = new AtomicInteger();
+
+        EventsHandler cb1 = new EventsHandlerCounter(count);
+        EventsHandler cb2 = new EventsHandlerCounter(count);
+        EventsHandler cberr = new EventsHandlerErrorCounter(error);
+        Pipe[] pipes = new Pipe[2];
+        try (
+             ZPoller poller = new ZPoller(selector)) {
+            pipes[0] = pipe(poller, cberr, cb1, cb2);
+            pipes[1] = pipe(poller, cberr, cb1, cb2);
+            int max = 10;
+            do {
+                poller.poll(100);
+                --max;
+            } while (count.get() != 4 && max > 0);
+
+            assertThat("Not all events handlers checked", error.get(), is(equalTo(0)));
+            if (max == 0 && count.get() != 4) {
+                fail("Unable to poll after 10 attempts");
+            }
+        }
+        finally {
+            selector.close();
+            for (Pipe pipe : pipes) {
+                pipe.sink().close();
+                pipe.source().close();
+            }
+        }
+    }
+
+    private Pipe pipe(ZPoller poller, EventsHandler errors, EventsHandler... ins) throws IOException
+    {
+        Pipe pipe = Pipe.open();
+        Pipe.SourceChannel source = pipe.source();
+        source.configureBlocking(false);
+        for (EventsHandler handler : ins) {
+            poller.register(source, handler, ZPoller.IN);
+        }
+        poller.register(source, errors, ZPoller.ERR);
+
+        pipe.sink().write(ByteBuffer.wrap(new byte[] { 1, 2, 3 }));
+
+        return pipe;
     }
 }

